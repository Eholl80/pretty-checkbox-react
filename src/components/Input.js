// @flow

import * as React from 'react';
import classNames from 'classnames';

/**
 * The base prefix for all pretty-checkbox class names.
 */
export const PREFIX: string = 'p-';

export type InputProps = {
    /**
     * Select the type if component: checkbox or radio.
     */
    type: "checkbox" | "radio",

    /**
     * Customize the rendering of the checkbox, radio,
     * or switch.
     *
     * **Note:** You are responsible for providing
     * the details of pretty-checkbox's `div.state`.
     */
    children?: React.Node | (() => React.Node),

    /**
     * Customize the rendering of the checkbox, radio,
     * or switch.
     *
     * **Note:** You are responsible for providing
     * the details of pretty-checkbox's `div.state`.
     */
    render?: (() => React.Node),

    /**
     * The style of the checkbox or radio.
     */
    style?: "fill" | "thick",

    /**
     * The shape of the checkbox or radio component.
     */
    shape?: "round" | "curve",

    /**
     * Render a custom font icon in the checkbox or radio.
     */
    icon?: React.Node,

    /**
     * Render a custom `.svg` in the checkbox or radio.
     */
    svg?: React.Element<'svg'>,

    /**
     * Render a custom `img` in the checkbox or radio.
     */
    image?: React.Element<'img'>,

    /**
     * Additional class selectors to pass to the `pretty` element.
     */
    className?: string,

    /**
     * Specify a value for the underlying `input` element.
     */
    value?: string | number | string[],

    /**
     * Add an event handler when the radio/checkbox/switch changes.
     */
    onChange?: (SyntheticMouseEvent<HTMLInputElement>) => void,

    /**
     * Need a11y? Add your `id` here and it'll automatically be added to the `label`
     * when you use `children` as a prop and not a function.
     */
    id?: string,

    /**
     * Additional input props to pass to the input component.
     */
    inputProps?: {},

    /**
     * Specify colors to add to the checkbox, radio, or switch.
     */
    color?: string,

    /**
     * Specify animations to add to the checkbox, radio, or switch.
     */
    animation?: "smooth" | "jelly" | "tada" | "rotate" | "pulse",

    /**
     * Control the state of your component by deciding when it can be checked/unchecked.
     */
    checked?: boolean,

    /**
     * Disable the checkbox, radio, or switch.
     */
    disabled?: boolean,

    /**
     * Lock the checkbox, radio, or switch.
     */
    locked?: boolean,

    /**
     * Make the checkbox, radio, or switch bigger.
     */
    bigger?: boolean,

    /**
     * Style the checkbox or radio as plain.
     */
    plain?: boolean
};

const fillClassNameForIcons = (component: React.Component<any>, className: string): React.Node => {
    if (!component) {
        return null;
    }

    return React.cloneElement(component, {
        ...component.props,
        className: classNames(className, component.props.className)
    });
};

/**
 * Handles custom or default rendering of the pretty-checkbox `div.state` class.
 */
const PrettyInputState = ({ children, render, id, color, icon, svg, image }: InputProps): React.Node => {
    if (typeof children === 'function') {
        return children();
    }

    if (typeof render === 'function') {
        return render();
    }

    return (
        <div className={classNames('state', color ? PREFIX + color : null)}>
            {fillClassNameForIcons(icon, 'icon') || fillClassNameForIcons(svg, 'svg') || fillClassNameForIcons(image, 'image') || null}
            {children ? <label htmlFor={id}>{children}</label> : null}
        </div>
    );
};

function Input(props: InputProps) {
    const {
        className,
        value,
        onChange,
        id,
        type,
        inputProps,
        animation,
        checked,
        disabled,
        locked,
        bigger,
        shape,
        style,
        image,
        svg,
        icon,
        plain
    } = props;

    if ((icon && svg) || (icon && image) || (svg && image)) {
        throw new Error('icon, svg, and image are mutually exclusive props; choose one');
    }

    return (
        <div
            className={classNames(
                'pretty',
                animation ? PREFIX + animation : null,
                className,
                shape ? PREFIX + shape : null,
                style ? PREFIX + style : null,
                locked ? `${PREFIX}locked` : null,
                bigger ? `${PREFIX}bigger` : null,
                plain ? `${PREFIX}plain` : null
            )}
        >
            <input
                id={id || null}
                type={type}
                value={value}
                onChange={onChange}
                checked={checked}
                disabled={disabled}
                {...inputProps}
            />
            <PrettyInputState {...props} />
        </div>
    );
}

export default Input;
