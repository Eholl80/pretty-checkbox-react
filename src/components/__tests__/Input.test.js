import React from 'react';

import { render, cleanup, getByTestId } from '@testing-library/react';
import 'jest-dom/extend-expect';

import Input from '../Input';

afterEach(cleanup);

describe('Input tests', function() {
    describe('Input custom rendering', function() {
        it('should support custom rendering via render prop and children render function', function() {
            const { container, rerender } = render(
                <Input type="checkbox">
                    {() => <p data-testid="custom-test">Hello</p>}
                </Input>
            );
            expect(() => getByTestId(container, 'custom-test')).not.toThrow();

            rerender(
                <Input
                    type="checkbox"
                    render={() => <p data-testid="custom-test">Hello</p>}
                />
            );
            expect(() => getByTestId(container, 'custom-test')).not.toThrow();
            expect(container).toMatchSnapshot();
        });
    });

    describe('Input errors', function() {
        it('should throw when svg, image, and icon are used in combination', function() {
            console.error = jest.fn(); // sshhhhhh

            expect(() =>
                render(
                    <Input icon={<i />} svg={<svg />}>
                        Hello there.
                    </Input>
                )
            ).toThrow();
        });
    });

    describe('Input ref tests', function() {
        it('should accept custom refs', function () {
            const ref = React.createRef();

            render(<Input ref={ref} />);
            expect(ref.current).toBeDefined();
            expect(ref.current).toBeInstanceOf(HTMLInputElement);
        });
    });

    describe('Input className auto prefix', function() {
        it('should prefix colors', function() {
            const { container } = render(
                <Input color="primary">Hello there.</Input>
            );

            expect(getByTestId(container, 'pcr-state')).toHaveClass(
                'p-primary'
            );
        });

        it('should prefix animations', function() {
            const { container } = render(
                <Input animation="smooth">Hello there.</Input>
            );

            expect(getByTestId(container, 'pcr-wrapper')).toHaveClass(
                'p-smooth'
            );
        });

        it('should prefix styles', function() {
            const { container } = render(
                <Input style="fill">Hello there.</Input>
            );

            expect(getByTestId(container, 'pcr-wrapper')).toHaveClass('p-fill');
        });

        it('should prefix utilities', function() {
            const { container } = render(
                <Input bigger locked>
                    Hello there.
                </Input>
            );

            expect(getByTestId(container, 'pcr-wrapper')).toHaveClass(
                'p-locked'
            );

            expect(getByTestId(container, 'pcr-wrapper')).toHaveClass(
                'p-bigger'
            );
        });

        it('should prefix plain', function() {
            const { container } = render(<Input plain>Hello there.</Input>);

            expect(getByTestId(container, 'pcr-wrapper')).toHaveClass(
                'p-plain'
            );
        });
    });
});
