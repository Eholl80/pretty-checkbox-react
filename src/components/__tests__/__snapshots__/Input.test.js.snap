// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input tests Input custom rendering should support custom rendering via render prop and children render function 1`] = `
<div>
  <div
    class="pretty"
    data-testid="pcr-wrapper"
  >
    <input
      data-testid="pcr-input"
      type="checkbox"
      value=""
    />
    <p
      data-testid="custom-test"
    >
      Hello
    </p>
  </div>
</div>
`;
